### ISSUE TITLE
TensorRT

### ISSUE THREAD

**Issue by Akababa** [cloveranon/Clover-Edition#99]
I'm trying to speed up inference on gpt2 with TensorRT. Someone claimed they got [3x performance](https://github.com/minimaxir/gpt-2-simple/issues/146) with torch2trt, and Nvidia also had good results with [TRT BERT](https://github.com/NVIDIA/TensorRT/tree/release/5.1/demo/BERT/python/). In addition, TRT optimization reduces the model file size. 

However, I ran into a couple obstacles:

- TRT only works with fixed-size inputs, so some masking inputs will be necessary (like how
[BERT](https://devblogs.nvidia.com/nlu-with-tensorrt-bert/) does it). Looking into the [GPT2 library docs](https://huggingface.co/transformers/model_doc/gpt2.html#gpt2lmheadmodel)
- Licensing issues with CUDA/Cudnn/TRT means it can't be easily distributed - although the users who already jumped through the hoops for cudnn probably won't mind doing the same for TRT if it gives another boost to inference speed.

Colab notebook showing how to install TensorRT and use torch2trt:
https://colab.research.google.com/drive/1oe_aflRfCwRehho_8QlD8YFKUQ_I5sA6

**Comment by AIDungeonWiXAnon** [comment]
Could be an interesting avenue to explore, although the whole distribution thing sounds like it might be a stickler. it's not an outright "don't share this with others" like cuDNN is it?
Also, have you experimented with masking inputs to the model and how the affects the output?
Regardless, anything we can do to get the model smaller and faster without castrating its performance is a good idea in my book.
Speaking of, is there a reason we split the model's weights to FP16 at runtime instead of distributing a FP16 version right off the bat @cloveranon?

Since it's in the same avenue, I've also been experimenting with training a KD'd version of GPT-2 and quantizing the model further from 32-bit --> 8-bit, or maybe even 2/3 bit if I can successfully implement the methodology outlined in [a recent paper](https://arxiv.org/pdf/1909.05840.pdf) and [one of its references](https://arxiv.org/pdf/1905.03696.pdf).

Knowledge Distillation off of the 1558M model, even with apex's FP16 O1 optimizations on a P100, is giving me memory capacity issues though; I hope to resolve this with parallelization assuming I can get a hold of an extra GPU or two, and quantization to extremely low precision still needs a lot of research on my end, as far as implementing the Hessian mean/variance math into usable code goes.
8-bit quantization is looking bright though, at least so far as making a non-Hessian based version to serve as a baseline for future tests.

**Comment by cloveranon** [comment]
We already use 16bit. I'm not sure we would gain anything from this. If I understand most of these gains are for training, not inference. Huggingface has a post that suggests there might be a 20% gain by using torchscript. And of course model distillation or further quantization is possible if anyone has the resources to do that.

If someone wants to host a 16 bit model by all means. I was hoping we would eventually have a whole collection of pytorch models available. However the models load into memory as 32 bit regardless of what you save them as.

**Comment by AIDungeonWiXAnon** [comment]
Well, at the very least hosting a 16-bit model means users only have a 3GB torrent to download and seed, rather than a 6GB one.
Even if we don't change anything else on the game's side for how it handles models, it'd be nice for all the poor end-users with slower connections. I already reconverted the original tensorflow model into a 16-bit pytorch model, so perhaps I could host that.

In the meantime I'll continue toying with further quantization and trying to get the GPUs needed for a distilled model.

**Comment by Akababa** [comment]
> We already use 16bit. I'm not sure we would gain anything from this. If I understand most of these gains are for training, not inference. 

I thought TRT was for inference, not training. Or do you mean most of the gains are from 16-bit so going the rest of the way with TRT won't help much?

> Could be an interesting avenue to explore, although the whole distribution thing sounds like it might be a stickler. it's not an outright "don't share this with others" like cuDNN is it?

I don't know but I went through more or less the same steps to obtain both, and they have the same use-case so I imagine the licenses don't differ too much. Here's the license for TRT:
https://developer.download.nvidia.com/compute/machine-learning/tensorrt/docs/TensorRT-SLA-Sep-2019.pdf

> Also, have you experimented with masking inputs to the model and how the affects the output?

I'm still figuring out how exactly gpt2 and the huggingface API works. I think you actually can't mask out the precomputed hidden states (`past` argument) so this might be a dead end, at least with huggingface. Nvidia's bert repo uses tons of custom code to get what they want.

**Comment by cloveranon** [comment]
I don't know anything about TRT specifically. What methods could they be using to get us that much gain? We are mostly constrained by memory size and bandwidth. I don't even think cuDNN stuff matters because all the optimizations assume multiple batches and we can only do 1.

And don't be limited by huggingface's api. All the code from their repo necessary to run GPT2 is about 400 lines.

**Comment by Akababa** [comment]
After much hair loss to TRT documentation (or lack thereof), I'm pretty close to getting this working. In the process I discovered some optimizations for the pure pytorch version that I hope to merge this week (and upstream to huggingface), and next week maybe I can put up a TRT engine optimized for colab.

**Comment by Damiox** [comment]
@Akababa have you already uploaded your fixes to huggingface's code? How did you compare small gpt2 ?

**Comment by salanki** [comment]
Any update on this?

**Comment by Akababa** [comment]
Sorry I haven't had the time to finish this. I did however merge a small optimization into transformers. https://github.com/huggingface/transformers/pull/2715

**Comment by jpark621** [comment]
Any updates as of November?

**Comment by ZhuoranLyu** [comment]
I am using onnxruntime-gpu to speed up the inference of GPT-2 model, from which I got a reasonable improvement compared with original pytorch or tensorflow.
